File: app/src/main/java/com/donald/musictheoryapp/Music/ScoreView/SSignature.java
package com.donald.musictheoryapp.Music.ScoreView;

import android.graphics.Canvas;

import com.donald.musictheoryapp.Music.MusicXML.Clef;
import com.donald.musictheoryapp.Music.MusicXML.Key;
import com.donald.musictheoryapp.Music.MusicXML.Note;

class SSignature extends ScoreStamp
{
	private static class SSignatureAccidental extends ScoreStamp
	{
		private static final int[] BASS_SHARP_POS = { 6, 3, 7, 4, 1, 5, 2 };
		private static final int[] BASS_FLAT_POS = { 2, 5, 1, 4, 0, 3, -1 };
		private static final int[] TENOR_SHARP_POS = { 2, 6, 3, 7, 4, 8, 5 };
		private static final int[] TENOR_FLAT_POS = { 5, 8, 4, 7, 3, 6, 2 };
		private static final int[] ALTO_SHARP_POS = { 7, 4, 8, 5, 2, 6, 3 };
		private static final int[] ALTO_FLAT_POS = { 3, 6, 2, 5, 1, 4, 0 };
		private static final int[] TREBLE_SHARP_POS = { 8, 5, 9, 6, 3, 7, 4 };
		private static final int[] TREBLE_FLAT_POS = { 4, 7, 3, 6, 2, 5, 1 };

		private char m_Glyph;
		private float m_VerticalOffsetOnStaff;
		private int[] m_CurrentAccidentalPositions;

		SSignatureAccidental(ScoreView view)
		{
			super(view);
			m_Glyph = U.ACC_NATURAL;
			m_VerticalOffsetOnStaff = 0;
		}

		protected void setKey(Key key, Clef clef)
		{
			boolean sharpSignature = key.fifths > 0;
			if(clef.sign == Clef.Sign.F && clef.line == 4)
			{
				// bass
				m_CurrentAccidentalPositions = sharpSignature ? BASS_SHARP_POS : BASS_FLAT_POS;
			}
			else if(clef.sign == Clef.Sign.C && clef.line == 3)
			{
				// tenor
				m_CurrentAccidentalPositions = sharpSignature ? TENOR_SHARP_POS : TENOR_FLAT_POS;
			}
			else if(clef.sign == Clef.Sign.C && clef.line == 4)
			{
				// alto
				m_CurrentAccidentalPositions = sharpSignature ? ALTO_SHARP_POS : ALTO_FLAT_POS;
			}
			else if(clef.sign == Clef.Sign.G && clef.line == 2)
			{
				// treble
				m_CurrentAccidentalPositions = sharpSignature ? TREBLE_SHARP_POS : TREBLE_FLAT_POS;
			}
		}

		protected void setAccidental(int nthAccidental, int alter)
		{
			m_VerticalOffsetOnStaff =
				(m_CurrentAccidentalPositions[nthAccidental])
				* staffStepHeight();
			switch(alter)
			{
				case -2:
					m_Glyph = U.ACC_D_FLAT;
					break;
				case -1:
					m_Glyph = U.ACC_FLAT;
					break;
				case 0:
					m_Glyph = U.ACC_NATURAL;
					break;
				case 1:
					m_Glyph = U.ACC_SHARP;
					break;
				case 2:
					m_Glyph = U.ACC_D_SHARP;
					break;
				default:
					throw new IllegalArgumentException(alter + " + is not a valid alter.");
			}
		}

		@Override
		protected void onDraw(Canvas canvas, float posX, float posY)
		{
			canvas.drawText(String.valueOf(m_Glyph),
				posX,
				posY + glyphSize() - m_VerticalOffsetOnStaff,
				glyphPaint()
			);
		}
	}

	private final SSignatureAccidental m_SAccidental;
	private Key m_Key;

	SSignature(ScoreView view)
	{
		super(view);
		m_SAccidental = new SSignatureAccidental(view);
	}

	protected void setKey(Key key, Clef clef)
	{
		m_Key = key;
		m_SAccidental.setKey(key, clef);
	}

	protected float getWidth()
	{
		if(m_Key == null) throw new IllegalStateException("Key is not set.");
		return Math.abs(m_Key.fifths) * noteWidth();
	}

	@Override
	protected void onDraw(Canvas canvas, float posX, float posY)
	{
		if(m_Key.fifths == 0) return;

		int firstStep;
		int factor;
		if(m_Key.fifths > 0)
		{
			// sharp signature
			firstStep = Note.Pitch.Step.STEPS_IN_FIFTHS[0];
			factor = 1;
			//signatureStartFifths = m_Key.fifths + 14;
			//nthNoteInSignature =
			//	Math.floorMod(Note.Pitch.Step.FIFTHS[firstStep] - signatureStartFifths,
			//		Note.Pitch.Step.NO_OF_STEPS);
		}
		else
		{
			// flat signature
			firstStep = Note.Pitch.Step.STEPS_IN_FIFTHS[Note.Pitch.Step.STEPS_IN_FIFTHS.length - 1];
			factor = -1;
			//signatureStartFifths = m_Key.fifths + 14;
			//nthNoteInSignature =
			//	Math.floorMod(Note.Pitch.Step.FIFTHS[firstStep] - signatureStartFifths,
			//		Note.Pitch.Step.NO_OF_STEPS);
		}
		// for iterating through the scale
		int signatureStartFifths = m_Key.fifths + 14;
		// initializing it to step F (i.e. Fbb, Fb, F, F#, or F##)
		int nthNoteInSignature = Math.floorMod(Note.Pitch.Step.FIFTHS[firstStep] - signatureStartFifths,
			Note.Pitch.Step.NO_OF_STEPS);

		// TODO: PERHAPS FIGURE OUT HOW TO PROCEDURALLY CALCULATE THE POSITIONS OF ACCIDENTALS
		int currentPosX = 0;
		for(int i = 0; i < 7; i++)
		{
			int currentFifths = Math.floorMod(nthNoteInSignature + i * factor, Note.Pitch.Step.NO_OF_STEPS) +
				signatureStartFifths;
			int currentAlter = currentFifths / Note.Pitch.Step.NO_OF_STEPS - 2;
			if(currentAlter == 0) break;
			m_SAccidental.setAccidental(i, currentFifths / Note.Pitch.Step.NO_OF_STEPS - 2);
			m_SAccidental.onDraw(canvas, posX + currentPosX, posY);
			currentPosX += noteWidth();
		}
	}

	/*
		int currentPosX = 0;
		for(int i = 0; i < 7; i++)
		{
			int currentFifths = (nthNoteInSignature + i) % Note.Pitch.Step.NO_OF_STEPS +
				signatureStartFifths;
			int currentAccidentalPosition =
				(nthPositionFromBottom + 4 * i) % 9 +
					bottomNotePosition;
			m_SAccidental.setAccidental(currentAccidentalPosition,
				currentFifths / Note.Pitch.Step.NO_OF_STEPS - 2, m_Clef);
			m_SAccidental.onDraw(canvas, posX + currentPosX, posY);
			currentPosX += noteWidth();
		}

		 */
}

File: app/src/main/java/com/donald/musictheoryapp/music/scoreview/SSignature.kt
package com.donald.musictheoryapp.music.scoreview

import android.graphics.Canvas
import com.donald.musictheoryapp.music.MusicXML.Clef
import com.donald.musictheoryapp.music.MusicXML.Key
import com.donald.musictheoryapp.music.MusicXML.Pitch
import com.google.android.material.math.MathUtils.floorMod
import java.lang.IllegalArgumentException
import kotlin.math.abs

internal class SSignature(view: ScoreView) : ScoreStamp(view) {

    internal class SSignatureAccidental(view: ScoreView) : ScoreStamp(view) {

        private var glyph: Char
        private var verticalOffsetOnStaff: Float
        private lateinit var currentAccidentalPositions: IntArray
        override val relContentWidth = REL_SINGLE_ACC_WIDTH // ignoring key signatures with double flats

        fun setKey(key: Key, clef: Clef) {
            val sharpSignature = key.fifths > 0
            if (clef.sign == Clef.Sign.F && clef.line == 4) {
                // bass
                currentAccidentalPositions = if (sharpSignature) BASS_SHARP_POS else BASS_FLAT_POS
            } else if (clef.sign == Clef.Sign.C && clef.line == 3) {
                // tenor
                currentAccidentalPositions = if (sharpSignature) TENOR_SHARP_POS else TENOR_FLAT_POS
            } else if (clef.sign == Clef.Sign.C && clef.line == 4) {
                // alto
                currentAccidentalPositions = if (sharpSignature) ALTO_SHARP_POS else ALTO_FLAT_POS
            } else if (clef.sign == Clef.Sign.G && clef.line == 2) {
                // treble
                currentAccidentalPositions = if (sharpSignature) TREBLE_SHARP_POS else TREBLE_FLAT_POS
            }
        }

        fun setAccidental(nthAccidental: Int, alter: Int) {
            verticalOffsetOnStaff = (currentAccidentalPositions[nthAccidental] * staffStepHeight)
            glyph = when (alter) {
                -2 -> U.ACC_D_FLAT
                -1 -> U.ACC_FLAT
                0 -> U.ACC_NATURAL
                1 -> U.ACC_SHARP
                2 -> U.ACC_D_SHARP
                else -> throw IllegalArgumentException("$alter + is not a valid alter.")
            }
        }

        override fun onDraw(canvas: Canvas, posX: Float, posY: Float) {
            canvas.drawText(
                glyph.toString(),
                posX,
                posY + glyphSize - verticalOffsetOnStaff,
                glyphPaint
            )
        }

        companion object {
            private val BASS_SHARP_POS = intArrayOf(6, 3, 7, 4, 1, 5, 2)
            private val BASS_FLAT_POS = intArrayOf(2, 5, 1, 4, 0, 3, -1)
            private val TENOR_SHARP_POS = intArrayOf(2, 6, 3, 7, 4, 8, 5)
            private val TENOR_FLAT_POS = intArrayOf(5, 8, 4, 7, 3, 6, 2)
            private val ALTO_SHARP_POS = intArrayOf(7, 4, 8, 5, 2, 6, 3)
            private val ALTO_FLAT_POS = intArrayOf(3, 6, 2, 5, 1, 4, 0)
            private val TREBLE_SHARP_POS = intArrayOf(8, 5, 9, 6, 3, 7, 4)
            private val TREBLE_FLAT_POS = intArrayOf(4, 7, 3, 6, 2, 5, 1)
        }

        init {
            glyph = U.ACC_NATURAL
            verticalOffsetOnStaff = 0f
        }
    }

    private val sAccidental: SSignatureAccidental = SSignatureAccidental(view)
    private lateinit var key: Key
    override val relContentWidth: Float
        get() = abs(key.fifths) * REL_SOLID_NOTE_WIDTH

    fun setKey(key: Key, clef: Clef) {
        this.key = key
        sAccidental.setKey(key, clef)
    }

    override fun onDraw(canvas: Canvas, posX: Float, posY: Float) {
        if (key.fifths == 0) return
        val firstStep: Int
        val factor: Int
        if (key.fifths > 0) {
            // sharp signature
            firstStep = Pitch.Step.STEPS_IN_FIFTHS[0]
            factor = 1
            //signatureStartFifths = m_Key.fifths + 14;
            //nthNoteInSignature =
            //	Math.floorMod(Note.Pitch.Step.FIFTHS[firstStep] - signatureStartFifths,
            //		Note.Pitch.Step.NO_OF_STEPS);
        } else {
            // flat signature
            firstStep = Pitch.Step.STEPS_IN_FIFTHS[Pitch.Step.STEPS_IN_FIFTHS.size - 1]
            factor = -1
            //signatureStartFifths = m_Key.fifths + 14;
            //nthNoteInSignature =
            //	Math.floorMod(Note.Pitch.Step.FIFTHS[firstStep] - signatureStartFifths,
            //		Note.Pitch.Step.NO_OF_STEPS);
        }
        // for iterating through the scale
        val signatureStartFifths = key.fifths + 14
        // initializing it to step F (i.e. Fbb, Fb, F, F#, or F##)
        val nthNoteInSignature = floorMod(
            Pitch.Step.FIFTHS[firstStep] - signatureStartFifths,
            Pitch.Step.NO_OF_STEPS
        )

        // TODO: PERHAPS FIGURE OUT HOW TO PROCEDURALLY CALCULATE THE POSITIONS OF ACCIDENTALS
        var currentPosX = 0
        for (i in 0..6) {
            val currentFifths = floorMod(nthNoteInSignature + i * factor, Pitch.Step.NO_OF_STEPS) +
                    signatureStartFifths
            val currentAlter = currentFifths / Pitch.Step.NO_OF_STEPS - 2
            if (currentAlter == 0) break
            sAccidental.setAccidental(i, currentFifths / Pitch.Step.NO_OF_STEPS - 2)
            sAccidental.onDraw(canvas, posX + currentPosX, posY)
            currentPosX += noteWidth.toInt()
        }
    }

}